include/oneapi/dpl/internal/random_impl/philox_engine.h

```diff
@@ -0,0 +1,533 @@
+
+    template <typename __CharT, typename __Traits, typename __UIntType, std::size_t __w, std::size_t __n,
+              std::size_t __r, internal::element_type_t<__UIntType>... __consts>
+    friend std::basic_istream<__CharT, __Traits>&
+    operator>>(std::basic_istream<__CharT, __Traits>&, philox_engine<__UIntType, __w, __n, __r, __consts...>&);
+
+  private:
+    /* Internal generator state */
+    struct state
+    {
+        std::array<scalar_type, word_count> X;     // counters
+        std::array<scalar_type, word_count / 2> K; // keys
+        std::array<scalar_type, word_count> Y;     // results
+        scalar_type idx;                           // index
+    } state_;
+
+    /* __word_mask<_W> - scalar_type with the low _W bits set */
+    template <std::size_t _W>
+    static constexpr scalar_type __word_mask = _W ? (~scalar_type(0) >> (std::numeric_limits<scalar_type>::digits - _W))
+                                                  : 0;
+
+    /* Processing mask */
+    static constexpr auto in_mask = __word_mask<word_size>;
+
+    void
+    seed_internal(scalar_type __seed)
+    {
+        // set to zero counters and results
+        for (std::size_t __i = 0; __i < word_count; ++__i)
+        {
+            state_.X[__i] = 0;
+            state_.Y[__i] = 0;
+        }
+        // 0th key element is set as seed, others are 0
+        state_.K[0] = __seed & in_mask;
+        for (std::size_t __i = 1; __i < (word_count / 2); ++__i)
+        {
+            state_.K[__i] = 0;
+        }
+
+        state_.idx = word_count - 1;
+    }
+
+    /* Increment counter by 1 */
+    void
+    increase_counter_internal()
+    {
+        for (std::size_t __i = 0; __i < word_count; ++__i)
+        {
+            state_.X[__i] = (state_.X[__i] + 1) & in_mask;
+            if (state_.X[__i])
+            {
+                return;
+            }
+        }
+    }
+
+    /* Increment counter by an arbitrary __z */
+    void
+    increase_counter_internal(unsigned long long __z)
+    {
+        unsigned long long __carry = 0;
+        unsigned long long __ctr_inc = __z;
+
+        for (std::size_t __i = 0; __i < word_count; ++__i)
+        {
+            scalar_type __initial_x = state_.X[__i];
+            state_.X[__i] = (__initial_x + __ctr_inc + __carry) & in_mask;
+
+            __carry = 0;
+            // check overflow of the current chunk
+            if (state_.X[__i] < __initial_x)
+            {
+                __carry = 1;
+            }
+
+            //          select high chunk             shift for addition with the next counter chunk
+            __ctr_inc = (__ctr_inc & (~in_mask)) >> (std::numeric_limits<unsigned long long>::digits - word_size);
+        }
+    }
+
+    /* generate_internal() specified for sycl_vec output 
+       and overload for result portion generation */
+    template <unsigned int _N = 0>
+    std::enable_if_t<(_N > 0), result_type>
```

Comment:
this enable_if can be removed as it is another overload

Reply:
Decided offline to leave it consistent with other engines.